File: src/test/java/org/opensearch/index/analysis/utils/AnalyzerFactory.java
package org.opensearch.index.analysis.utils;

import org.apache.lucene.analysis.Analyzer;
import org.apache.lucene.analysis.TokenStream;
import org.apache.lucene.analysis.Tokenizer;
import org.apache.lucene.analysis.core.KeywordTokenizer;
import org.opensearch.index.analysis.chosung.ChosungTokenFilter;
import org.opensearch.index.analysis.engtokor.ConvertEngToKoreanFilter;
import org.opensearch.index.analysis.jamo.JamoDecomposeTokenFilter;
import org.opensearch.index.analysis.kortoeng.ConvertKoreanToEngFilter;
import org.opensearch.index.analysis.spell.SpellFilter;

import java.io.IOException;

public abstract class AnalyzerFactory {

    private AnalyzerFactory() {
        throw new IllegalArgumentException("It's Util Class. Do not create instance");
    }

    public static Analyzer createAnalyzer(AnalyzerType type) {
        return new Analyzer(Analyzer.PER_FIELD_REUSE_STRATEGY) {
            @Override
            protected TokenStreamComponents createComponents(String fieldName) {
                try (Tokenizer tokenizer = new KeywordTokenizer()) {
                    TokenStream tokenFilter = switch (type) {
                        case CHOSUNG_TOKEN -> new ChosungTokenFilter(tokenizer);
                        case JAMO_DEOMPOSE -> new JamoDecomposeTokenFilter(tokenizer);
                        case CONVERT_ENG_TO_KOREAN -> new ConvertEngToKoreanFilter(tokenizer);
                        case CONVERT_KOREAN_TO_ENG -> new ConvertKoreanToEngFilter(tokenizer);
                        case SPELL_CHECK -> new SpellFilter(tokenizer);
                    };

                    return new TokenStreamComponents(tokenizer, tokenFilter);
                } catch (IOException ex) {
                    throw new RuntimeException(ex);
                }
            }
        };
    }
}

File: src/test/java/org/opensearch/index/analysis/utils/AnalyzerFactory.kt
package org.opensearch.index.analysis.utils

import org.apache.lucene.analysis.Analyzer
import org.apache.lucene.analysis.TokenFilter
import org.apache.lucene.analysis.TokenStream
import org.apache.lucene.analysis.core.KeywordTokenizer
import java.io.IOException

/*
* Licensed to Elasticsearch B.V. under one or more contributor
* license agreements. See the NOTICE file distributed with
* this work for additional information regarding copyright
* ownership. Elasticsearch B.V. licenses this file to you under
* the Apache License, Version 2.0 (the "License"); you may
* not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*      http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing,
* software distributed under the License is distributed on an
* "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
* KIND, either express or implied. See the License for the
* specific language governing permissions and limitations
* under the License.
*/
object AnalyzerFactory {

    fun <T : TokenFilter> createAnalyzer(constructor: (TokenStream) -> T): Analyzer =
        object: Analyzer(PER_FIELD_REUSE_STRATEGY) {
            override fun createComponents(fieldName: String): TokenStreamComponents =
                try {
                    KeywordTokenizer().use { tokenizer ->
                        TokenStreamComponents(tokenizer, constructor(tokenizer))
                    }
                } catch (ex: IOException) {
                    throw RuntimeException(ex)
                }
        }
    }
